### PR TITLE
Changes to support forward references.

### PR DESCRIPTION
**Pull request by bader**
This change is required for enabling forward reference support used list data structure.

struct Node {
    Node *next;
    int value;
};

### CODE REVIEW

File: lib/SPIRV/libSPIRV/SPIRVType.cpp [hunk 1]
```diff
@@ -240,6 +240,27 @@ SPIRVType::isTypeVectorOrScalarFloat() const {
...
+void SPIRVTypeStruct::encode(spv_ostream &O) const {
+  getEncoder(O) << Id << MemberTypeVec;
+}
+
+void SPIRVTypeStruct::decode(std::istream &I) {
+  auto Decoder = getDecoder(I);
+  Decoder >> Id;
+
+  for (size_t i = 0, e = MemberTypeVec.size(); i != e; ++i) {
+    SPIRVId currId;
```
Comment (yxsamliu) [inline lib/SPIRV/libSPIRV/SPIRVType.cpp]: currId should be capitalized for LLVM style, i and e may be renamed as MI and ME

File: lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -380,6 +380,37 @@ LLVMToSPIRV::isBuiltinTransToExtInst(Function *F,
   return true;
 }
 
+static bool recursiveType(const StructType *ST, const Type *Ty) {
+  SmallPtrSet<const StructType *, 4> Seen;
+
+  std::function<bool(const Type *Ty)> run = [&](const Type *Ty) {
```
Comment (yxsamliu) [inline lib/SPIRV/SPIRVWriter.cpp]: run should be capitalized for LLVM style

File: test/SPIRV/transcoding/RecursiveType.ll [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+; RUN: llvm-spirv %t.bc -spirv-text -o %t.txt
+; RUN: FileCheck < %t.txt %s --check-prefix=CHECK-SPIRV
+; RUN: llvm-spirv %t.bc -o %t.spv
+; RUN: llvm-spirv -r %t.spv -o %t.rev.bc
+; RUN: llvm-dis < %t.rev.bc | FileCheck %s --check-prefix=CHECK-LLVM
+
+target datalayout = "e-p:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024"
+target triple = "spir-unknown-unknown"
+
+%struct.Node = type { %struct.Node addrspace(1)*, i32 }
```
Comment (yxsamliu) [inline test/SPIRV/transcoding/RecursiveType.ll:14]: Better to add a test for nested structs, e.g.

struct A;
struct B {
  int b1;
  struct A *b2;
};
struct C {
  int c1;
  struct B c2;
};
struct A {
  int a1;
  struct C a2;
};
global struct A a;

the IR is:
%struct.A = type { i32, %struct.C }
%struct.C = type { i32, %struct.B }
%struct.B = type { i32, %struct.A\* }

@a = common global %struct.A zeroinitializer, align 8

Comment (bader) [inline test/SPIRV/transcoding/RecursiveType.ll:14]: Added similar case to the test.
Also I've added support for OpTypeForwardPointer as it's required by SPIR-V specification: https://www.khronos.org/registry/spir-v/specs/1.0/SPIRV.html#_universal_validation_rules (data rules).

**Comment by yxsamliu** [comment]
LGTM. Please resolve conflicts.

**Comment by bader** [comment]
Merged with the latest version of khronos/spirv-3.6.1 branch.

**Pull request merged**